db/product.go

```diff
@@ -0,0 +1,313 @@
+	Brand        string   `db:"brand" json:"brand"`
+	Color        string   `db:"color" json:"color"`
+	Size         string   `db:"size" json:"size"`
+	URLs         []string `json:"image_url,omitempty"`
+}
+
+// Pagination helps to return UI side with number of pages given a limit and page
+type Pagination struct {
+	Products   []Product `json:"products"`
+	TotalPages int       `json:"total_pages"`
+}
+
+func (product *Product) Validate() (errorResponse map[string]ErrorResponse, valid bool) {
+	fieldErrors := make(map[string]string)
+
+	if product.Name == "" {
+		fieldErrors["product_name"] = "Can't be blank"
+	}
+	if product.Description == "" {
+		fieldErrors["product_description"] = "Can't be blank "
+	}
+	if product.Price <= 0 {
+		fieldErrors["price"] = "Can't be blank  or less than zero"
+	}
+	if product.Discount < 0 {
+		fieldErrors["discount"] = "Can't be less than zero"
+	}
+	if product.Tax < 0 {
+		fieldErrors["tax"] = "Can't be less than zero"
+	}
+	// If Quantity gets's < 0 by UpdateProductStockById Method, this is what saves us
+	if product.Quantity < 0 {
+		fieldErrors["available_quantity"] = "Can't be blank or less than zero"
+	}
+	if product.CategoryId == 0 {
+		fieldErrors["category_id"] = "Can't be blank"
+	}
+
+	if len(fieldErrors) == 0 {
+		valid = true
+		return
+	}
+
+	errorResponse = map[string]ErrorResponse{
+		"error": ErrorResponse{
+			Code:    "Invalid_data",
+			Message: "Please Provide valid Product data",
+			Fields:  fieldErrors,
+		},
+	}
+	// TODO Other Validations
+	return
+}
+
+// @Title GetProductByID
+// @Description Get a Product Object by its Id
+// @Params req.Context, product's Id
+// @Returns Product Object, error if any
+func (s *pgStore) GetProductByID(ctx context.Context, Id int) (product Product, err error) {
+
+	err = s.db.Get(&product, getProductByIDQuery, Id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting product from database by id: " + string(Id))
+		return
+	}
+
+	// Add category to Product's object
+	var category string
+	err = s.db.Get(&category, getCategoryByID, product.CategoryId)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error fetching category from database by product_id: " + string(Id))
+		return
+	}
+
+	product.CategoryName = category
+
+	// Add Product Image URL's to Product's Object
+	productImage, err := s.GetProductImagesByID(ctx, Id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting productImage from database by id " + string(Id))
+		return
+	}
+
+	for j := 0; j < len(productImage); j++ {
+		product.URLs = append(product.URLs, productImage[j].URL)
+	}
+
+	return
+}
+
+// @Title ListProducts
+// @Description Get limited number of Products of particular page
+// @Params req.Context , limit, page
+// @Returns Count of Records, error if any
+func (s *pgStore) ListProducts(ctx context.Context, limit string, page string) (count int, products []Product, err error) {
```

Comment:
should we use named return params for long methods?
I kinda get confused at each return

Reply:
Specifying return values at each place will make the return statement look monster, but we will do it.

Reply:
Done